WatchFaceAlphaKotlin/app/src/main/java/com/example/android/wearable/alpha/AnalogWatchCanvasRenderer.kt

```diff
@@ -0,0 +1,546 @@
+            val dy = -cos(rotation).toFloat() * numberRadiusFraction * bounds.width().toFloat()
+            textPaint.getTextBounds(HOUR_MARKS[i], 0, HOUR_MARKS[i].length, textBounds)
+            canvas.drawText(
+                HOUR_MARKS[i],
+                bounds.exactCenterX() + dx - textBounds.width() / 2.0f,
+                bounds.exactCenterY() + dy + textBounds.height() / 2.0f,
+                textPaint
+            )
+        }
+
+        // Draws dots for the remain hour indicators between the numbers above.
+        outerElementPaint.strokeWidth = outerCircleStokeWidthFraction * bounds.width()
+        outerElementPaint.color = outerElementColor
+        canvas.save()
+        for (i in 0 until 12) {
+            if (i % 3 != 0) {
+                drawTopMiddleCircle(
+                    canvas,
+                    bounds,
+                    numberStyleOuterCircleRadiusFraction,
+                    gapBetweenOuterCircleAndBorderFraction
+                )
+            }
+            canvas.rotate(360.0f / 12.0f, bounds.exactCenterX(), bounds.exactCenterY())
+        }
+        canvas.restore()
+    }
+
+    /** Draws the outer circle on the top middle of the given bounds. */
+    private fun drawTopMiddleCircle(
+        canvas: Canvas,
+        bounds: Rect,
+        radiusFraction: Float,
+        gapBetweenOuterCircleAndBorderFraction: Float
+    ) {
+        outerElementPaint.style = Paint.Style.FILL_AND_STROKE
+
+        // X and Y coordinates of the center of the circle.
+        val centerX = 0.5f * bounds.width().toFloat()
+        val centerY = bounds.width() * (gapBetweenOuterCircleAndBorderFraction + radiusFraction)
+
+        canvas.drawCircle(
+            centerX,
+            centerY,
+            radiusFraction * bounds.width(),
+            outerElementPaint
+        )
+    }
+
+    // Saves user style changes to the database.
+    suspend fun updateUserStylesInDatabase(userStyle: UserStyle) {
+        val analogWatchFaceEntity: AnalogWatchFaceEntity =
+            analogWatchFaceViewModel.getAnalogWatchFace(analogWatchFaceKeyId)
+
+        updateColorStyleAndHourPips(
+            userStyle,
+            analogWatchFaceEntity
+        )
+
+        updateMinuteArmLength(
+            userStyle,
+            analogWatchFaceEntity.minuteHandDimensionsId
+        )
+    }
+
+    // Updates the color style and hour hand pips (ticks on the outside of the watch) in the
+    // associated [AnalogWatchFaceEntity] in the database.
+    private suspend fun updateColorStyleAndHourPips(
+        userStyle: UserStyle,
+        analogWatchFaceEntity: AnalogWatchFaceEntity
+    ) {
+        var revisedAnalogWatchFaceEntity = analogWatchFaceEntity
+
+        // 1. Updates color style associated with the watch face.
+        val newColorStyleSetting: String =
+            userStyle.toMap()[COLOR_STYLE_SETTING] ?: revisedAnalogWatchFaceEntity.activeColorStyleId
```

Comment:
FYI we're adding some helpers to make UserStyle a bit nicer, see: https://android-review.googlesource.com/c/platform/frameworks/support/+/1556081 That won't land till in a release for a few weeks though.

The way I was anticipating UserStyle working is using the UserStyleSetting as the key to the map although this certainly works.

Reply:
I will check that out and add it back to this.